micropip/package_index.py

```diff
@@ -0,0 +1,121 @@
+    size: int | None = None
+
+
+@dataclass
+class ProjectInfo:
+    """
+    This class stores common metadata that can be obtained from different APIs (JSON, Simple)
+    provided by PyPI. Responses received from PyPI or other package indexes that support the
+    same APIs must be converted to this class before being processed by micropip.
+    """
+
+    name: str  # Name of the package
+
+    # List of releases available for the package, sorted in ascending order by version
+    # This list only contains wheels not sdist, but it does not filter by tags for it might contain
+    # binary wheels for other platforms. So it is up to the caller to filter the list by tags.
+    # TODO: should we filter by tags here? Or should we leave it to the caller?
+    releases: dict[Version, list[ProjectInfoFile]]
+
+    @staticmethod
+    def from_json_api(data: dict[str, Any]) -> "ProjectInfo":
+        """
+        Parse JSON API response
+
+        https://warehouse.pypa.io/api-reference/json.html
+        """
+
+        name: str = data.get("info", {}).get("name", "UNKNOWN")
+        _releases: dict[str, Any] = data["releases"]
+
+        releases: dict[Version, list[ProjectInfoFile]] = defaultdict(list)
+        for version_str, fileinfo in _releases.items():
+            try:
+                version = Version(version_str)
+            except InvalidVersion:
+                # Ignore non PEP 440 compliant versions
+                continue
+
+            for file in fileinfo:
+                filename = file["filename"]
+                if not filename.endswith(".whl"):
+                    continue
+
+                releases[version].append(
+                    ProjectInfoFile(
+                        filename=filename,
+                        url=file["url"],
+                        version=version,
+                        sha256=file["digests"]["sha256"],
+                        size=file["size"] if "size" in file else None,
+                    )
+                )
+
+        # Unfortunately, the JSON API seems to compare versions as strings...
+        # For example, pytest 3.10.0 is considered newer than 3.2.0.
+        # So we need to sort the releases by version again here.
+        releases = dict(sorted(releases.items()))
+
+        return ProjectInfo(
+            name=name,
+            releases=releases,
+        )
+
+    @staticmethod
+    def from_simple_api(data: dict[str, Any]) -> "ProjectInfo":
+        """
+        Parse Simple API response
+
+        https://peps.python.org/pep-0503/
+        https://peps.python.org/pep-0691/
+        """
+
+        name = data["name"]
+        releases: defaultdict[Version, list[ProjectInfoFile]] = defaultdict(list)
+        for file in data["files"]:
+            filename = file["filename"]
+            if not filename.endswith(".whl"):
+                continue
+
+            try:
+                version = parse_wheel_filename(filename)[1]
```

Comment:
A bit unfortunate to parse the wheel filename twice once here, and once when filtering out by platform, but it's fast enough, so shouldn't matter.
```
In [2]: %timeit parse_wheel_filename('scikit_learn-1.3.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl')
6.06 µs ± 31.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```